### DISCUSSION TITLE
Cross compiling to mac failed by thread-local

### DISCUSSION THREAD

**Discussion by addsoa** [libusb/libusb#1585] [Q&A] [answered]
I am cross-compiling to mac arm by
```  bash
    cd $LIBUSB_BUILD_PATH; \
        rm -rf *; \
        CFLAGS="$COMMON_CFLAGS" \
        LDFLAGS="$COMMON_CLINKFLAGS" \
        $LIBUSB_SRC_PATH/configure \
        --prefix=$LIBUSB_TARGET_PATH \
        --host=$TARGET_TRIPLE \
        --disable-silent-rules \
        --with-pic \
        --disable-tests \
        --enable-static \
        --with-sysroot=$SYSROOT_PATH \
        --with-gnu-ld=no 

    cd $LIBUSB_BUILD_PATH; make VERBOSE=1; make install
```

The output:
``` c
libtool: compile:  clang -DHAVE_CONFIG_H -I. -I/home/me/workspaces/plain_sources/libusb/libusb -I.. -std=gnu11 -Wall -Wextra -Wshadow -Wunused -Wwrite-strings -Werror=format-security -Werror=implicit-function-declaration -Werror=implicit-int -Werror=init-self -Werror=missing-prototypes -Werror=strict-prototypes -Werror=undef -Werror=uninitialized -fvisibility=hidden -pthread -fPIC -std=c11 -fvisibility=hidden -fvisibility-inlines-hidden --target=arm64-apple-darwin-mach --sysroot=/home/me/images/Xcode15_4_SDK/ -MT os/threads_posix.lo -MD -MP -MF os/.deps/threads_posix.Tpo -c /home/me/workspaces/plain_sources/libusb/libusb/os/threads_posix.c  -fno-common -DPIC -o os/.libs/threads_posix.o
/home/me/workspaces/plain_sources/libusb/libusb/os/threads_posix.c:86:9: error: thread-local storage is not supported for the current target
   86 |         static _Thread_local unsigned int tl_tid;
      |                ^
1 error generated.
make[1]: *** [Makefile:657: os/threads_posix.lo] Error 1
```

I've raised an [issue to llvm](https://github.com/llvm/llvm-project/issues/116562).

Is there any workaround so I can go on by cross-compiling?

Thank you very much!

**Comment by seanm** [comment]
What version of libusb? what version of macOS? what version of clang? etc.

**Reply by addsoa** [reply]
libusb: libussb-1.0.26
clang: clang17, clang18
macos: Xcode15.4

**Reply by seanm** [reply]
When you say "cross compiling to mac" do you mean your trying to build *for* macOS from a different OS?

**Reply by addsoa** [reply]
Yes. I want to compile for macOS arm64 on ubuntu x86_64.

**Reply by mcuee** [reply, marked as answer]
Maybe the toolchain is broken?

It is better to have a Mac when you need to develop for Mac. I usually get the cheapest Mac Mini myself, Mac Mini 2011 (upgraded RAM to 8GB) initially in 2011 and then Mac Mini M1 (8GB/256GB version) in 2020. 

That being said, for avrdude project, we also use the Arduino Cross-build container to cross build for Windows, Linux and macOS for avrdude and its dependencies like libusb, hidapi and libftdi.

**Reply by addsoa** [reply]
Yes, the toolchain is broken, so I am looking for some workarounds in source codes.  While it is just discussion, I've built it on a cheap Mac. Thank you very much for you advices. 

I know nothing about Arduino. So, Arduino Cross-build container is a good choice for libusb cross-building?

**Reply by mcuee** [reply]
Arduino cross-build container has nothing to do with Arduino in reality. You can try it to see if it fits your use case or not.
https://github.com/arduino/crossbuild

It has built-in support of the following library but you can use it to cross-build other libraries, or newer version of the libraries.
https://github.com/arduino/crossbuild/tree/main/deps

I am also the admin of avrdude project. In avrdude project, we use it to cross build avrdude binaries for Linux, macOS and Windows.
https://github.com/avrdudes/avrdude/blob/main/.github/workflows/arduino_packing_release.yml

**Reply by addsoa** [reply]
README of `arduino` says it is for `mac_amd64` not for `mac_arm64`. Seemed a mac as builder is more reliable solution.  Thank you very much for your help.
